File: Go/1 Arrays/Q18.go
// The program below finds two array elements whose sum is closest to zero.

package main

import (
	"fmt"
	"math"
)

func findMinSumPair(arr []int) (int, int, bool) {
	if len(arr) < 2 {
		return 0, 0, false // Not enough elements to find a pair
	}

	min1Pair := arr[0]
	min2Pair := arr[1]
	minSum := min1Pair + min2Pair

	for i := 0; i < len(arr)-1; i++ {
		for j := i + 1; j < len(arr); j++ {
			sum := arr[i] + arr[j]
			if math.Abs(float64(sum)) < math.Abs(float64(minSum)) {
				minSum = sum
				min1Pair = arr[i]
				min2Pair = arr[j]
			}
		}
	}

	return min1Pair, min2Pair, true
}

func main() {
	arr := []int{10, 1, 24, -10, -2, 19, 22, -24, 4, -1}

	fmt.Println("The original array:", arr)

	// Find and display the pair with a sum closest to zero
	if a, b, found := findMinSumPair(arr); found {
		fmt.Printf("The pair of elements whose sum is minimum are: %d and %d\n", a, b)
	} else {
		fmt.Println("The array does not contain enough elements to find a pair.")
	}
}

File: Rust/1 Arrays/Q18.rs
// The program below finds two array elements whose sum is closest to zero.


fn find_min_sum_pair(arr: &[i32]) -> Option<(i32, i32)> {
    // Check if the array is valid and has at least two elements
    if arr.len() < 2 {
        return None;
    }

    let mut min1_pair = arr[0];
    let mut min2_pair = arr[1];
    let mut min_sum = min1_pair + min2_pair;


    for i in 0..arr.len() - 1 {     // Find the pair of elements with the minimum sum closest to zero
        for j in i + 1..arr.len() {
            let sum = arr[i] + arr[j];
            if sum.abs() < min_sum.abs() {
                min_sum = sum;
                min1_pair = arr[i];
                min2_pair = arr[j];
            }
        }
    }

    Some((min1_pair, min2_pair))
}

fn main() {
    let arr = vec![9, 7, 22, -10, -2, 19, 84, -77, 4, -1];

    println!("The original array: {:?}", arr);

    match find_min_sum_pair(&arr) {
        Some((a, b)) => {
            println!("The pair of elements whose sum is minimum are: {} and {}", a, b);
        }
        None => {
            println!("The array does not contain enough elements to find a pair.");
        }
    }
}